### <center>Midterm Term Exam </center>

<center>(Due on Friday, March 27)
   
    
    
In this exam, you are asked to read the paper on Batch Normalization which is available on BBLearn, and write python code in the $k$-hidden layer model. Then, use the model to classify the symbols in the test set in BBLearn. The training set is the same as the one in this year's Data Science Competition avaiable at https://dscomp.nkn.uidaho.edu/
        
Here I briefly describe batch normalization: Suppose $O_1$, $O_2$, ..., are the output of the $i$-th layer before applying activation function $a$,  
$$ O_1= XW_1+B_1;\ \ O_i=a(O_{i-1})W_i+B_i,$$ 
for $i=2,\ldots, n$. 

Suppose $O_{i-1}$ is an $N\times n_i$ matrix. Before applying activation function, we normalize each column vector of $O_{i-1}$ so that the new column vector will have a desired mean and standard deviation. This is done by first  subtracting its mean and devided it by its standard deviation, so that the new column vector will have mean 0 and standard deviation 1. More precisely, if the $j$-th column vector is vector $o_j^{(i-1)}$. The corresponding  new column is then
$$\widetilde{o}_j^{(i-1)}=\frac{o_j^{(i-1)}-np.mean(o_j^{(i-1)})}{\sqrt{np.var(o_j^{(i-1)})+\varepsilon}},$$
where $\varepsilon$ is a small positive number called smooth factor to prevent from dividing by 0. Denote 
<span style="color:blue">
$$\widetilde{O}^{(i-1)}=(\widetilde{o}_ j^{(i-1)}).$$
</span>

While the above step is probably not much inovation, the batch normalization proposed by Ioffe and Szegedy in 2015 takes further a creative step: scaling and shifting $\widetilde{o}_j^{(i-1)}$'s using learnable parameters!

$$\widehat{o}_j^{(i-1)}= \gamma_j^{(i-1)}\cdot \widetilde{o}_j^{(i-1)}+\beta_j^{(i-1)}=\gamma_j^{(i-1)}\cdot \frac{v_j^{(i-1)}-np.mean(o_j^{(i-1)})}{\sqrt{np.var(o_j^{(i-1)})+\varepsilon}}+\beta_j^{(i-1)}. $$
Thus, if we let 
<span style="color:blue">
$$\widehat{O}^{(i-1)}=(\widehat{o}_j^{(i-1)})=\Gamma^{(i-1)}\otimes \widetilde{O}^{(i-1)}+\beta^{(i-1)},$$
</span>
where 
$$\Gamma^{(i-1)}=\begin{bmatrix}\gamma_1^{(i-1)}&\gamma_2^{(i-1)}&\cdots&\gamma_{n_i}^{(i-1)}\\
    \gamma_1^{(i-1)}&\gamma_2^{(i-1)}&\cdots&\gamma_{n_i}^{(i-1)}\\
    \vdots&\vdots&\cdots&\vdots\\
    \gamma_1^{(i-1)}&\gamma_2^{(i-1)}&\cdots&\gamma_{n_i}^{(i-1)}
\end{bmatrix}=\begin{bmatrix}1\\1\\\vdots\\1
\end{bmatrix}\begin{bmatrix}\gamma_1^{(i-1)}&\gamma_2^{(i-1)}&\cdots&\gamma_{n_i}^{(i-1)}
\end{bmatrix},$$
$$\beta^{(i-1)}=\begin{bmatrix}\beta_1^{(i-1)}&\beta_2^{(i-1)}&\cdots&\beta_{n_i}^{(i-1)}\\
    \beta_1^{(i-1)}&\beta_2^{(i-1)}&\cdots&\beta_{n_i}^{(i-1)}\\
    \vdots&\vdots&\cdots&\vdots\\
    \beta_1^{(i-1)}&\beta_2^{(i-1)}&\cdots&\beta_{n_i}^{(i-1)}
\end{bmatrix}=\begin{bmatrix}1\\1\\\vdots\\1
\end{bmatrix}\begin{bmatrix}\beta_1^{(i-1)}&\beta_2^{(i-1)}&\cdots&\beta_{n_i}^{(i-1)}
\end{bmatrix},$$ we can write
<span style="color:blue">
$$ O_1=XW_1,\ \ \ \ O_i=a\left(\widehat{O}^{(i-1)}\right)W_i, \ i=2,3,\ldots, n.$$ 
</span>
(Note: there is no need for the original biases $B_i$.) Because of these learnable scaling and shifting parameteres $\gamma_{i-1}$ and $\beta_{i-1}$, neural network performance is less dependent on weights initialization and regularization. Note that using numpy default, we can code $\Gamma^{(i-1)}=[[\gamma_1^{(i-1)}, \gamma_2^{(i-1)},\ldots, \gamma_{n_i}^{(i-1)}]]$ and $\beta^{(i-1)}=[[\beta_1^{(i-1)}, \beta_2^{(i-1)},\ldots, \beta_{n_i}^{(i-1)}]]$ as shape $1\times n_i$. However, when doing gradient calculation, we need to know that the exact expression is $$\Gamma^{(i-1)}={\mathbb 1}[[\gamma_1^{(i-1)}, \gamma_2^{(i-1)},\ldots, \gamma_{n_i}^{(i-1)}]],$$
$$\beta^{(i-1)}={\mathbb 1}[[\beta_1^{(i-1)}, \beta_2^{(i-1)},\ldots, \beta_{n_i}^{(i-1)}]],$$
    where ${\mathbb 1}=np.ones((N,n_i))$,

The rest of the neural network may stay the same.
    
Recall that the main gradient formulas
  
$\frac{\partial L}{\partial U}=\frac{\partial L}{\partial a(U)}\otimes (U>0)$  
    
$\frac{\partial L}{\partial U}=\frac{\partial L}{\partial (U\otimes A)}\otimes A$  
    
$\frac{\partial L}{\partial U}=\frac{\partial L}{\partial (U+A)}$

$\frac{\partial L}{\partial U}=A^T\frac{\partial L}{\partial (AU)}$
  
$\frac{\partial L}{\partial U}=\frac{\partial L}{\partial (UA)}A^T$
    
From these formulas, we can obtain for example
   
$\frac{\partial L}{\partial W_i}=a(\widehat{O}^{(i-1)})^T\frac{\partial L}{\partial O_i}$
    
$\frac{\partial L}{\partial \widehat{O}^{(i-1)}}=\frac{\partial L}{\partial a(\widehat{O}^{(i-1)})}\otimes (\widehat{O}^{(i-1)}>0)=\left(\frac{\partial L}{\partial O_{i}}W_i^T\right)\otimes (\widehat{O}^{(i-1)}>0)$

Using the second expression above, you can get $\frac{\partial L}{\partial \Gamma^{(i-1)}}$ and $\frac{\partial L}{\partial \beta^{(i-1)}}$. 
    
Finish the codes below, and then use it to classify the attached testset. The exam will be graded by both the correctness of the coding and prediction accuracy. 
    
Optional: You may receive extra credits if you also implement other activation functions, or other optimization methods.
    


## Function to Subidivide a Dataset into Training and Testing Data

In [194]:
import numpy as np # Basic number package
import traceback   # File-traceback-tracing
import pickle      # Storing object instances to file
np.set_printoptions(precision=3)

def subdivide_data(x, y, test_percentage=0.025):
    assert len(x) == len(y), "Lengths of x and y must be equal."
    indices = np.arange(len(x))
    np.random.shuffle(indices)
    
    return {"X-Train": x[indices][:int(len(x) * (1.0 - test_percentage))], "Y-Train": y[indices][:int(len(x) * (1.0 - test_percentage))],
            "X-Test": x[indices][-int(len(x) * test_percentage):], "Y-Test": y[indices][-int(len(x) * test_percentage):]}

## Definition of my Activation Function

### I am using a Leaky ReLU Function for this network

$$a(x)=\left\{
\begin{array}{ll}
    0.1x & x\leq 0 \\
    x & x>0
\end{array}
\right.$$

In [181]:
activation_function = lambda x: x if x > 0 else 0.1 * x
back_activation_function = lambda x: 1 if x > 0 else 0.1

activation_function = np.vectorize(activation_function)
back_activation_function = np.vectorize(back_activation_function)

## K-Hidden Layer Neural Network Class

### I added layer-by-layer dropout to the Network (the second parameter during the `init`)

In [124]:
class k_hidden_layer_bn(object):
    def __init__(self, sizes, dropouts=None):
        self.object_name = traceback.extract_stack()[-2][3][:traceback.extract_stack()[-2][3].find('=')].strip() # Name of the instance of this class object
        self.num_layers = len(sizes)        
        self.weights = [np.random.randn(m, n) / np.sqrt(m) for m, n in zip(sizes[:-1], sizes[1:])]
        self.gammas = [np.ones((1, size)) for size in sizes[1:-1]] # Batch Normalization 'weight'
        self.betas = [np.zeros((1, size)) for size in sizes[1:-1]] # Batch Normalization 'bias'
        if dropouts == None:
            self.dropouts = [1.0 for layer in range(len(sizes))]
        else:
            self.dropouts = dropouts
        
    def predict(self, X, v=False):
        if v:
            print (self.loss(X))
        return np.argmax(self.loss(X), axis=1)
        
    def loss(self, X, y=None, reg=0.0):
        if y is None: # When predicting, don't dropout any nodes
            output = np.dot(X, self.weights[0])
        else:
            output = np.dot(X, self.weights[0] * ((np.random.uniform(0, 1.0, (self.weights[0].shape[0], 1)) > self.dropouts[0]) * 1.0))
        layer_out_list, cache_list, y_i_list = [output], [], []
        # Inputs -> Weights -> [Normalization -> Activation -> Weights] -> Output
        for i in range(self.num_layers - 2):
            x_min_mu = output - np.mean(output, axis=0, keepdims=True)
            variance = np.var(output, axis=0, keepdims=True) + 1e-8
            x_hat = x_min_mu / np.sqrt(variance)
            y_i = x_hat * self.gammas[i] + self.betas[i]
            
            if y is None: # When predicting, don't implement dropouts
                output = np.dot(activation_function(y_i), self.weights[i + 1])
            else:
                output = np.dot(activation_function(y_i), self.weights[i + 1] * ((np.random.uniform(0, 1.0, (self.weights[i + 1].shape[0], 1)) > self.dropouts[i + 1]) * 1.0))
            layer_out_list.append(output); cache_list.append((x_hat, x_min_mu, variance)); y_i_list.append(y_i)
            
        if y is None: # Good
            return np.exp(output - np.max(output)) / np.max(output)

        N = X.shape[0]
        P = np.zeros_like(output)
        P[range(N), y] = 1.0
        M = np.max(output, axis=1, keepdims=True)
        Q = np.exp(output - M) / np.sum(np.exp(output - M), axis=1, keepdims=True)
        loss = -np.mean(np.log(Q[range(N), y] + 1e-300))
        for W in self.weights:
            loss += reg * np.sum(W ** 2)

        d_gammas = [np.zeros_like(gamma) for gamma in self.gammas]
        d_betas = [np.zeros_like(beta) for beta in self.betas]
        d_outs = [np.zeros_like(o) for o in layer_out_list]
        
        d_outs[-1] = (Q - P) / N
        for i in range(2, self.num_layers):
            x_hat, x_min_mu, variance = cache_list[-i + 1]
            dl_dy = back_activation_function(np.dot(d_outs[-i + 1], self.weights[-i + 1].T)) # Back propagate through (activation <- Weights <- Output[next stage])
            
            # Intermediate variables for computing the back propogations
            dxhat = dl_dy * np.ones((N, self.gammas[-i + 1].shape[1])) * self.gammas[-i + 1]
            dsqrtvar = -1 / variance * np.sum(dxhat * x_min_mu, axis=0)
            dsq = np.ones((N, self.gammas[-i + 1].shape[1])) / N * 0.5 / np.sqrt(variance) * dsqrtvar
            dxmu2 = 2 * x_min_mu * dsq
            dmu = -np.sum((dxhat / variance) + dxmu2, axis=0)
            dx2 = np.ones((N, self.gammas[-i + 1].shape[1])) * dmu / N
            
            # Compute the back-propogation of dL/dy, dL/dgamma, and dL/dbeta
            d_outs[-i] = (dxhat / variance) + dxmu2 + dx2
            d_gammas[-i + 1] = np.sum(dl_dy * x_hat, axis=0)
            d_betas[-i + 1] = np.sum(dl_dy, axis=0)
            
        # Update the weights based on the back-propagated dL/dy values calculated above
        d_weights = [np.zeros_like(W) for W in self.weights]
        d_weights[0] = np.dot(X.T, d_outs[0]) + 2 * reg * self.weights[0]
        for i in range(1, self.num_layers - 1):
            d_weights[i] = np.dot((y_i_list[i - 1]).T, d_outs[i]) + 2 * reg * self.weights[i]

        return loss, d_weights, d_gammas, d_betas
    
    def train(self, X_train, y_train, X_val=None, y_val=None, learning_rate=1e-3, learning_rate_decay=0.99, reg=5e-3, epochs=5, batch_size=50, verbose=False):
        loss_history, train_acc_history, val_acc_history = [], [], []
        N = X_train.shape[0]
        iteration = 0
        
        for epoch in range(epochs):
            perm = np.arange(N)
            np.random.shuffle(perm)
            X_batches = [X_train[perm[k:k+batch_size]] for k in range(0, N, batch_size)]
            y_batches = [y_train[perm[k:k+batch_size]] for k in range(0, N, batch_size)]
            for X, y in zip(X_batches, y_batches):
                L, dWs, dgammas, dbetas = self.loss(X, y, reg)
                self.weights = [W - learning_rate * dW for W, dW in zip(self.weights, dWs)]
                self.gammas = [gamma - learning_rate * dgamma for gamma, dgamma in zip(self.gammas, dgammas)]
                self.betas = [beta - learning_rate * dbeta for beta, dbeta in zip(self.betas, dbetas)]
                
                if iteration % 50 == 0:  
                    loss_history.append(L)   
                    y_pred = self.predict(X)
                    train_acc = np.mean(y_pred == y)
                    train_acc_history.append(train_acc)
                    
                    if X_val is not None:
                        y_val_pred = self.predict(X_val)
                        val_acc = np.mean(y_val_pred == y_val) 
                        val_acc_history.append(val_acc)
                        
                    if verbose and X_val is not None:
                        print('\n\riteration %d: loss %f, train_acc: %f, val_acc: %f'  % (iteration, L, train_acc, val_acc), end='')
                    elif verbose:
                        print('\n\riteration %d: loss %f, train_acc: %f'% (iteration, L, train_acc), end='')
                            
                iteration+=1
                
            if X_val is not None:
                print('\nEpoch ', epoch+1, 'validation accuracy: ', val_acc)
            else:
                print('\nEpoch ', epoch+1, 'completed.')
                
            learning_rate *= learning_rate_decay
            
        return {'loss_history': loss_history, 'train_acc_history': train_acc_history, 'val_acc_history': val_acc_history}
    
    def save_to_file(self, name=None):
        filename = '{}.pickle'.format(self.object_name) if name is None else '{}.pickle'.format(name)
        file = open(filename, 'wb')
        pickle.dump(self, file)
        file.close()

## Load the Datasets

In [7]:
train_data = np.load('/Volumes/Seagate Backup Plus Drive/Miscellaneous/DataScience2020-Data/train_data.npy').reshape(8000, -1)
train_labels = np.load('/Volumes/Seagate Backup Plus Drive/Miscellaneous/DataScience2020-Data/train_labels.npy')
test_data = np.loadtxt('test_data.npy', delimiter=',')

## Train the Network

In [156]:
train_dict = subdivide_data(train_data, train_labels, 0.05)
network = k_hidden_layer_bn([4096, 2048, 16], [0.1, 0.25, 0.1])
history_dict = network.train(train_dict['X-Train'], train_dict['Y-Train'], train_dict['X-Test'], train_dict['Y-Test'],
     epochs=16, batch_size=64, verbose=True, learning_rate=5e-4, learning_rate_decay=0.6)


iteration 0: loss 13.119193, train_acc: 0.953125, val_acc: 0.460000
iteration 100: loss 17.798836, train_acc: 0.765625, val_acc: 0.702500
Epoch  1 validation accuracy:  0.7025

iteration 200: loss 17.582096, train_acc: 0.843750, val_acc: 0.717500
Epoch  2 validation accuracy:  0.7175

iteration 300: loss 18.165887, train_acc: 0.843750, val_acc: 0.730000
Epoch  3 validation accuracy:  0.73

iteration 400: loss 18.590751, train_acc: 0.718750, val_acc: 0.732500
Epoch  4 validation accuracy:  0.7325

iteration 500: loss 18.909615, train_acc: 0.734375, val_acc: 0.727500
Epoch  5 validation accuracy:  0.7275

iteration 600: loss 18.886856, train_acc: 0.781250, val_acc: 0.740000
iteration 700: loss 19.106629, train_acc: 0.765625, val_acc: 0.732500
Epoch  6 validation accuracy:  0.7325

iteration 800: loss 19.274452, train_acc: 0.750000, val_acc: 0.735000
Epoch  7 validation accuracy:  0.735

iteration 900: loss 19.228977, train_acc: 0.734375, val_acc: 0.737500
Epoch  8 validation accuracy:  

The final training accuracy I was able to achieve with this model was 81.25% after 16 epochs, with a valiation accuracy of 73.75%; however if I had trained the network for longer I suspect I could improve this metric.

## Make Predictions on the test data

### The results are saved to `predictions.csv`

In [180]:
np.savetxt('predictions.csv', [network.predict(val)[0] for val in test_data], fmt='%u', delimiter=',')
network.save_to_file()